File: JavaScript/Day12.js
(function(){
	"use strict";
	
	require("fs").readFile("../inputs/Day12.txt", "utf8", function(err, data){
		if(err){
			console.log(`Unable to load file: ${err}`);
		} else {
			data = JSON.parse(data);
			let allNumbers = countNumbers(data);
			let noRedNumbers = countNumbers(data, true);

			console.log(`Part 1: ${allNumbers}`);
			console.log(`Part 2: ${noRedNumbers}`);
		}
	});

	let countNumbers = function(object, skipReds){
		skipReds = skipReds || false;

		if(skipReds && typeof object === 'object' && !Array.isArray(object)){
			for(let k in object){
				if(object[k] === 'red'){
					return 0;
				}
			}
		}
		
		let total = 0;
		for(let key in object){
			if(typeof object[key] === 'object'){
				total += countNumbers(object[key], skipReds);
			} else if(typeof object[key] == 'number'){
				total += object[key];
			}
		}
		return total;
	}
})();

File: Python/Day12.py
import json

data = open('../inputs/Day12.txt').read()
data = json.loads(data)

def countNumbers(obj, skipReds = False):
	total = 0

	if isinstance(obj, int):
		total += obj
	elif isinstance(obj, list):
		total += sum( countNumbers(o, skipReds) for o in obj )
	elif isinstance(obj, dict):
		if skipReds and 'red' in obj.values():
			total = 0
		else:
			total += sum( countNumbers(o, skipReds) for o in obj.values() )

	return total

allNumbers = countNumbers(data)
noRedNumbers = countNumbers(data, True)

print "Part 1:", allNumbers
print "Part 2:", noRedNumbers